<h1><center>Laboratorio 3: El Pandas no muerde 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos - Primavera 2024</strong></center>

### Cuerpo Docente:

- Profesores: Ignacio Meza, Sebastián Tinoco
- Auxiliar: Eduardo Moya
- Ayudantes: Nicolás Ojeda, Melanie Peña, Valentina Rojas

### Equipo: **SUPER IMPORTANTE - notebooks sin nombre no serán revisados**

- Nombre de alumno 1: Carolina Nuñez
- Nombre de alumno 2: Alonso Uribe


### **Link de repositorio de GitHub:** [Repositorio💻](https://github.com/carinunez/Labs_MDS/)

## Reglas:

- **Grupos de 2 personas**
- Fecha de entrega: 6 días de plazo con descuento de 1 punto por día. Entregas Jueves a las 23:59.
- Instrucciones del lab el viernes a las 16:15 en formato online. Asistencia no es obligatoria.
- <u>Prohibidas las copias</u>. Cualquier intento de copia será debidamente penalizado con el reglamento de la escuela.
- Tienen que subir el laboratorio a u-cursos y a su repositorio de github. Labs que no estén en u-cursos no serán revisados. Recuerden que el repositorio también tiene nota.
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Pueden usar cualquier material del curso que estimen conveniente.

### Temas a tratar

- Introducción al manejo de datos tabulares por medio de la libreria `pandas`.

### Objetivos principales del laboratorio

- Entender, aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas` para manejar datos tabulares.
- Aprender a utilizar la documentación de pandas (y en general, de cualquier librería de python) como también a buscar soluciones en la web.

> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

### Librerias utiles para el Lab 😸

In [1]:
# Libreria Core del lab.
import numpy as np
import pandas as pd
from IPython.display import display

#Libreria para plotear
#%pip install --upgrade plotly
import plotly.express as px

# Librerias utiles
from sklearn.preprocessing import StandardScaler

# 1. Rendimiento en Estudiantes 📚

Para este laboratorio deberán trabajar con el dataset `students_grades.json` el cual contiene información sobre el rendimiento de alumnos en distintos ramos.
El objetivo principal será el de realizar un trabajo de limpieza de prepración y exploración de datos utilizando exclusivamente `DataFrames` y sus métodos asociados.




In [2]:
# Si usted está utilizando Colabolatory le puede ser útil este código para cargar los archivos.
try:
    from google.colab import drive
    drive.mount("/content/drive")
    path = 'Dirección donde tiene los archivos en el Drive'
except:
    print('Ignorando conexión drive-colab')

Ignorando conexión drive-colab


## 1. Carga e Inspección de Datos [1.0 Puntos]


<p align="center">
  <img src="https://i.imgur.com/44knnIf.png" width="500" height="500">
</p>

*Don Caguayo*, rector de la Universidad de la Cachaña, les comenta que actualmente la Universidad no tiene una subdirección de análisis de datos, pero que tiene un gran interés en formar una. Es por esto, que se contacta con ud. y su equipo para iniciar de forma exploratoria diversas tareas de análisis y consolidación de datos respectivos al rendimiento de sus alumnos.

Previo a su primera reunión con el rector, el rector le envía una copia de la base de datos, la cuál a usted y su equipo le encantaría examinar. Para esto, la primera tarea que debe realizar es abrir y cargar el dataset.


> **Notar que es sumamente importante el formato que posee cada columna** para su manipulación y que posiblemente ustedes deberán especificar algunos tipos de datos a mano. Ver la documentación del método `read_json` para mayor información.

A continuación se muestra un ejemplo de los dos primeros registros del dataset:

```json
[
  {
    "names": "Rita Courtney",
    "gender": "female",
    "race/ethnicity": "group B",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 37,
    "reading score": 46,
    "writing score": 46
  },
  {
    "names": "Madeline Fuller",
    "gender": "female",
    "race/ethnicity": "group C",
    "parental level of education": "some high school",
    "lunch": "standard",
    "test preparation course": "none",
    "math score": 124,
    "reading score": 142,
    "writing score": 142
  }
]
```





In [3]:
df_grades = pd.read_json('students_grades.json')
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1      Madeline Fuller  female        group C            some high school   
2        Daniel Rester    male        group E          associate's degree   
3     Charles Linstrom    male        group A           bachelor's degree   
4       Geraldine Hale  female        group A          associate's degree   
...                ...     ...            ...                         ...   
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
0         standard                    none         37            46   
1         standard                    none        124           142   
2         standard                    none        156           196   
3         standard               completed         80            78   
4         standard                    none        131           166   
...            ...                     ...        ...           ...   
1395      standard                    none        193           142   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1398  free/reduced                    none        193           196   
1399      standard               completed         68            80   

     writing score  
0               46  
1              142  
2              193  
3               81  
4              166  
...            ...  
1395           193  
1396            81  
1397            62  
1398           193  
1399            76  

[1400 rows x 9 columns]

In [4]:
df_grades1 = pd.read_json('students_grades_1.json')
df_grades2 = pd.read_json('students_grades_2.json')
df_grades2 # Ambos datasets están asociados con df_grades, grades1 tiene 600 alumnos y grades2 400

names  gender race/ethnicity parental level of education  \
601   Anastasia Starks  female        group B                 high school   
602   Dolores Sprinkle  female        group C           bachelor's degree   
603       Jeremy Smith    male        group D            some high school   
604    Stella Stephens  female        group D             master's degree   
605         Paul Hyder    male        group D             master's degree   
...                ...     ...            ...                         ...   
998      Richard Young    male        group D                 high school   
999      Wanda Russell  female        group B                 high school   
1000    Marina Zeigler  female        group C           bachelor's degree   
1001     Laurie Carter  female        group B            some high school   
1002      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
601   free/reduced               completed        5.02           5.68   
602       standard                    none        5.02           5.14   
603       standard               completed        5.44           5.26   
604       standard                    none        5.44           5.74   
605       standard                    none        6.70           5.86   
...            ...                     ...         ...            ...   
998       standard                    none        5.14           5.50   
999   free/reduced               completed        2.38           3.64   
1000  free/reduced               completed        4.96           5.44   
1001      standard               completed        4.24           4.66   
1002      standard               completed        5.08           5.80   

      writing score   GPA GPA cuantiles  
601            5.74  5.48          Alto  
602            5.50  5.22         Medio  
603            5.68  5.46          Alto  
604            5.92  5.70          Alto  
605            6.04  6.20      Muy alto  
...             ...   ...           ...  
998            5.26  5.30         Medio  
999            3.16  3.06      Bajísimo  
1000           5.86  5.42          Alto  
1001           4.72  4.54          Bajo  
1002           5.56  5.48          Alto  

[400 rows x 11 columns]


<p align="center">
  <img src="https://i.giphy.com/media/4FQMuOKR6zQRO/giphy.webp">
</p>


Posteriormente, el rector les comenta que la sección de TI posee una base de datos con más de **1000** filas, en donde cada fila representa un alumno matriculado. Sin embargo, cree fuertemente que existen errores en esta: solo tiene registro de **1000** matriculas. Por este motivo, especula que se han realizado malas prácticas al registrar dichos datos.

En consecuencia, el rector les solicita a ustedes como tarea inicial que exploren la base de datos y encuentre posibles falencias dentro de sus registros. Para esto, usted debe estar al tanto que las notas están en una escala que van del **0 al 100** (porcentaje de logro) y **no es posible que un alumno haya dado más de una vez el mismo examen**. Cualquier registro que posea un valor diferente a esta escala, **debería ser eliminado de la base de datos**.


> **Instrucciones**

**Cree una función** que tome un `DataFrame` y que le permita revisar si las diferentes variables que lo componen son coherentes y si es posible observar anomalías en estas.

**To-Do**
1. [ ] Imprimir un mensaje con las dimensiones del Dataframe.
2. [ ] Imprimir un mensaje con los nombres de las columnas que conforman el DataFrame.
3. [ ] Mostrar los primeros 5, los últimos 5 elementos del DataFrame y realice un muestreo de 5 elementos de forma aleatoria.
4. [ ] Mostrar una descripción rápida de las variables numéricas del DataFrame (utilice un método visto en clases).
5. [ ] Imprimir un mensaje con la cantidad de valores nulos (investigue el método `isna()` y `sum()` de los DataFrames).
6. [ ] Imprimir un mensaje que detalle la cantidad de valores unicos (investigue `nunique()`) .
7. [ ] Imprimir un mensaje con el total de filas que contengan nombres duplicados (investigue `duplicated()` y su parámetro `subset`).
8. [ ] Mostrar las filas que contengan nombres sean duplicados (investigue `duplicated()` y su parámetro `subset`).

Finalmente, responda las siguientes preguntas (comente esto en la celda ubicada debajo de **justificación**.):

9. [ ] ¿Qué anomalías es posible observar de los datos que componen el DataFrame? ¿Existen valores nulos o duplicados?


10. [ ] Su compañero ChatGPT le provee un código para que pueda entender los resultados de las pruebas desde un gráfico. ¿Qué puede interpretar de la distribución de las notas en el boxplot?

> **Nota:** Para imprmir utilice `print(...)`. Para mostrar `DataFrames`, utilice `display(...)` ya que este permite mostrar estos elementos como tablas.

> **Nota 2:** Utilice la documentación de [pandas](https://pandas.pydata.org/docs/index.html) como también StackOverflow para resolver cualquier problema o inconveniente.

> **Nota 3:** Si no ha trabajado con boxplot antes, el siguiente link puede serle de mucha ayuda: [boxplot](https://medium.com/dayem-siddiqui/understanding-and-interpreting-box-plots-d07aab9d1b6c)

**Respuesta:**

In [5]:
# df_grades[df_grades.filter(like='score').columns] = df_grades[df_grades.filter(like='score').columns].apply(pd.to_numeric)

In [6]:
df_grades.loc[df_grades.duplicated(subset='names', keep='first')].sort_values('names')

names  gender race/ethnicity parental level of education  \
441       Aaron Oakden    male        group D                 high school   
1393       Adam Romero    male        group C                 high school   
1395  Alexander Dillon    male        group D          associate's degree   
1049     Amanda Barker  female        group C                some college   
1174         Amy Hawes  female        group B           bachelor's degree   
...                ...     ...            ...                         ...   
768   William Gardiner    male        group E                 high school   
1121        William Ng    male        group C            some high school   
1023     William Saska    male        group B          associate's degree   
476        Willie Neff  female        group E          associate's degree   
1355     Zachary Glenn    male        group E                some college   

             lunch test preparation course math score reading score  \
441   free/reduced                    none                      193   
1393      standard               completed        124           193   
1395      standard                    none        193           142   
1049  free/reduced               completed         64            85   
1174      standard                    none        142           147   
...            ...                     ...        ...           ...   
768   free/reduced                    none         55            56   
1121      standard                    none         62            64   
1023      standard                    none        193            42   
476   free/reduced                    none                      193   
1355      standard               completed         99            87   

     writing score  
441            124  
1393            49  
1395           193  
1049            85  
1174           193  
...            ...  
768             51  
1121            55  
1023           142  
476            142  
1355            81  

[247 rows x 9 columns]

In [7]:
df_grades.duplicated(subset='names').astype(int).sum()

247

In [8]:
def exploratory_data_analysis(dataframe):
    # Imprimir dim DataFrame
    print('Dimensiones del dataframe: ',dataframe.shape)
    print('------------------------------------------------------')
    # Imprimir nombre cols
    print('Columnas: ', list(dataframe.columns))
    print('------------------------------------------------------')

    # Mostrar head(5) tail(5) y 5 datos random
    print('Primeros 5 elementos')
    display(dataframe.head(5))

    print('Ultimos 5 elementos')
    display(dataframe.tail(5))
    print('5 elementos aleatorios')
    display(dataframe.sample(5))

    print('------------------------------------------------------')
    # mostrar descripción variables .describe()
    
    # Dado que las columnas que tienen variables numéricas de interés son de tipo 
    # objeto, antes de cambiar su formato, es necesario buscar las columnas que contengan 
    # la palabra 'score' en su nombre. 
    print('Descripción de las variables numéricas')
    cols_score =  dataframe.filter(like='score').columns
    dataframe[cols_score] = dataframe[cols_score].apply(pd.to_numeric)
    
    display(dataframe.describe())

    print('------------------------------------------------------')
    # imprimir num valores nulos -> isna().sum()
    print('Valores nulos')
    print(dataframe.isna().sum())

    print('------------------------------------------------------')
    # imprimir num valores unicos -> -nunique()
    print('Valores unicos')
    print(dataframe.nunique())

    print('------------------------------------------------------')
    # imprimir total filas con valores duplicados .duplicated(subset= ALGO)
    print('Total de filas duplicadas')
    print(dataframe.duplicated(subset='names').astype(int).sum())

    print('------------------------------------------------------')
    print('Nombres duplicados')
    display(dataframe.loc[dataframe.duplicated(subset='names', keep='first')].sort_values('names'))
    pass


In [9]:
# Código del amigo ChatGPT (listo para usar)
fig = px.box(
    df_grades.melt(
        value_vars=["math score", "reading score", "writing score"],
        var_name="subject",
        value_name="score",
    ),
    y="score",
    color="subject",
)
fig.show()

In [10]:
exploratory_data_analysis(df_grades)

Dimensiones del dataframe:  (1400, 9)
------------------------------------------------------
Columnas:  ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']
------------------------------------------------------
Primeros 5 elementos


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
1   Madeline Fuller  female        group C            some high school   
2     Daniel Rester    male        group E          associate's degree   
3  Charles Linstrom    male        group A           bachelor's degree   
4    Geraldine Hale  female        group A          associate's degree   

      lunch test preparation course math score reading score writing score  
0  standard                    none         37            46            46  
1  standard                    none        124           142           142  
2  standard                    none        156           196           193  
3  standard               completed         80            78            81  
4  standard                    none        131           166           166

Ultimos 5 elementos


names  gender race/ethnicity parental level of education  \
1395  Alexander Dillon    male        group D          associate's degree   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1398      Joseph Mccoy    male        group D                some college   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course math score reading score  \
1395      standard                    none        193           142   
1396  free/reduced               completed         66            74   
1397      standard               completed         54            61   
1398  free/reduced                    none        193           196   
1399      standard               completed         68            80   

     writing score  
1395           193  
1396            81  
1397            62  
1398           193  
1399            76

5 elementos aleatorios


names  gender race/ethnicity parental level of education  \
781      Monica Duggins  female        group B          associate's degree   
511      Robert Miranda    male        group C                 high school   
311      Gerald Brunner    male        group D                 high school   
683      James Crawford    male        group D             master's degree   
338  Florencia Morrison  female        group A          associate's degree   

            lunch test preparation course math score reading score  \
781      standard               completed         94            87   
511  free/reduced                    none        124           156   
311      standard                    none         46            34   
683      standard                    none         89            84   
338  free/reduced                    none         37            57   

    writing score  
781            92  
511           196  
311            36  
683            82  
338            56

------------------------------------------------------
Descripción de las variables numéricas


math score  reading score  writing score
count  1360.000000    1360.000000    1353.000000
mean     89.538971      93.086029      91.852919
std      47.152422      46.385775      47.043570
min       0.000000      17.000000      10.000000
25%      59.000000      63.000000      62.000000
50%      72.000000      76.000000      75.000000
75%      97.250000     100.000000     100.000000
max     196.000000     196.000000     196.000000

------------------------------------------------------
Valores nulos
names                           0
gender                          0
race/ethnicity                  0
parental level of education     0
lunch                           0
test preparation course         0
math score                     40
reading score                  40
writing score                  47
dtype: int64
------------------------------------------------------
Valores unicos
names                          1153
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       96
reading score                    88
writing score                    92
dtype: int64
------------------------------------------------------
Total de filas duplicadas
247
------------------------------------------------------
Nombres duplicados


names  gender race/ethnicity parental level of education  \
441       Aaron Oakden    male        group D                 high school   
1393       Adam Romero    male        group C                 high school   
1395  Alexander Dillon    male        group D          associate's degree   
1049     Amanda Barker  female        group C                some college   
1174         Amy Hawes  female        group B           bachelor's degree   
...                ...     ...            ...                         ...   
768   William Gardiner    male        group E                 high school   
1121        William Ng    male        group C            some high school   
1023     William Saska    male        group B          associate's degree   
476        Willie Neff  female        group E          associate's degree   
1355     Zachary Glenn    male        group E                some college   

             lunch test preparation course  math score  reading score  \
441   free/reduced                    none         NaN          193.0   
1393      standard               completed       124.0          193.0   
1395      standard                    none       193.0          142.0   
1049  free/reduced               completed        64.0           85.0   
1174      standard                    none       142.0          147.0   
...            ...                     ...         ...            ...   
768   free/reduced                    none        55.0           56.0   
1121      standard                    none        62.0           64.0   
1023      standard                    none       193.0           42.0   
476   free/reduced                    none         NaN          193.0   
1355      standard               completed        99.0           87.0   

      writing score  
441           124.0  
1393           49.0  
1395          193.0  
1049           85.0  
1174          193.0  
...             ...  
768            51.0  
1121           55.0  
1023          142.0  
476           142.0  
1355           81.0  

[247 rows x 9 columns]

**Justificación:**

**9. Anomalías en los datos**

Al observar los resultados, es posible notar que hay entre 40 y 50 datos faltantes en las columnas de notas. Además hay 247 nombres que se repiten. 

**10. Distribución de las notas** 

Las notas tienen medianas similares, con una diferencia indisitinguible. Al comparar los extremos de los blox plots se puede decir que:
- Matemáticas: es la única asignatura con un outlier en 0 y su percentil 75 es ligeramente menor al de las otras asignaturas. 
- Lectura: tiene una distribución más compacta (mínimos más cercano al percentil 25 que las otras asignaturas)
- Escritura: su máximo supera el alcanzado en las otras asignaturas (sin considerar outliers)

## 2 Limpieza de Datos [1.0 puntos]




<p align="center">
  <img src="https://blogdatlas.files.wordpress.com/2020/07/datlas_mx_blog_limpieza_bds_data-scrubbing-service.gif">
</p>

En base a lo realizado anteriormente, realice una limpieza de los dataset. Asuma los supuestos señalados en la pregunta anterior para limpiar el `DataFrame` y **cree una función** que le permita limpiar los datos. Habiendo limpiado el `DataFrame`, compruebe nuevamente los datos con la función creada en la sección 1.1.

**To-Do:**
- [ ] Eliminar registros con valores nulos.
- [ ] Eliminar valores fuera del rango válido.
- [ ] Eliminar registros con nombres duplicados
- [ ] Comprobar que no existen anomalías en los datos a traves de la función de la sección 1.1.

**Hints**

- Puede ser útil los métodos ```.drop_duplicates()``` y ```.dropna()``` como también utilizar queries con máscaras booleanas.
- Cuidado con el orden en la limpieza de datos. ¿Que se debe eliminar primero, los nombres duplicados o los registros fuera del rango posible?


> Utilice la función creada en el punto 1 para verificar sus resultados.

**Respuesta:**

In [11]:
df_grades.loc[(df_grades['math score']>100)|(df_grades['math score']<0)|
                (df_grades['reading score']>100)|(df_grades['reading score']<0)|
                (df_grades['writing score']>100)|(df_grades['writing score']<0)].index



Index([   1,    2,    4,   10,   11,   12,   14,   18,   22,   23,
       ...
       1346, 1350, 1358, 1369, 1378, 1385, 1390, 1393, 1395, 1398],
      dtype='int64', length=393)

In [12]:
# Escriba su código aquí
def clean_data(dataframe):
    # Eliminar registros con valores nulos.
    dataframe.dropna(inplace=True)

    # Eliminar valores fuera del rango válido.
    cols_score =  dataframe.filter(like='score').columns
    
    for col in cols_score:
        index_del = dataframe.loc[(dataframe[col]>100)|(dataframe[col]<0)].index
        dataframe.drop(index=index_del, axis=0, inplace=True)

    # Eliminar registros con nombres duplicados
    dataframe.drop_duplicates(subset='names', keep='first', inplace=True)
    
    return dataframe

In [13]:
#Pruebe su código aquí
df_grades = clean_data(df_grades)
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        37.0           46.0   
3         standard               completed        80.0           78.0   
5         standard                    none        73.0           66.0   
6         standard               completed        81.0           75.0   
7         standard               completed        94.0           86.0   
...            ...                     ...         ...            ...   
1392      standard                    none        69.0           75.0   
1394  free/reduced               completed        23.0           44.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1399      standard               completed        68.0           80.0   

      writing score  
0              46.0  
3              81.0  
5              63.0  
6              76.0  
7              87.0  
...             ...  
1392           71.0  
1394           36.0  
1396           81.0  
1397           62.0  
1399           76.0  

[1000 rows x 9 columns]

In [14]:
exploratory_data_analysis(df_grades)

Dimensiones del dataframe:  (1000, 9)
------------------------------------------------------
Columnas:  ['names', 'gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score']
------------------------------------------------------
Primeros 5 elementos


names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  
0  standard                    none        37.0           46.0           46.0  
3  standard               completed        80.0           78.0           81.0  
5  standard                    none        73.0           66.0           63.0  
6  standard               completed        81.0           75.0           76.0  
7  standard               completed        94.0           86.0           87.0

Ultimos 5 elementos


names  gender race/ethnicity parental level of education  \
1392   Richard Young    male        group D                 high school   
1394   Wanda Russell  female        group B                 high school   
1396  Marina Zeigler  female        group C           bachelor's degree   
1397   Laurie Carter  female        group B            some high school   
1399    Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
1392      standard                    none        69.0           75.0   
1394  free/reduced               completed        23.0           44.0   
1396  free/reduced               completed        66.0           74.0   
1397      standard               completed        54.0           61.0   
1399      standard               completed        68.0           80.0   

      writing score  
1392           71.0  
1394           36.0  
1396           81.0  
1397           62.0  
1399           76.0

5 elementos aleatorios


names  gender race/ethnicity parental level of education  \
1307   Peter Perez    male        group D             master's degree   
842   Ofelia Evans  female        group C             master's degree   
1053    John Ponce    male        group B           bachelor's degree   
508   Darron Allen    male        group C                 high school   
181   Audrey Darby  female        group D            some high school   

             lunch test preparation course  math score  reading score  \
1307      standard                    none        81.0           81.0   
842       standard               completed        81.0           91.0   
1053  free/reduced                    none        88.0           75.0   
508       standard                    none        61.0           56.0   
181       standard               completed        66.0           78.0   

      writing score  
1307           84.0  
842            87.0  
1053           76.0  
508            55.0  
181            78.0

------------------------------------------------------
Descripción de las variables numéricas


math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

------------------------------------------------------
Valores nulos
names                          0
gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64
------------------------------------------------------
Valores unicos
names                          1000
gender                            2
race/ethnicity                    5
parental level of education       6
lunch                             2
test preparation course           2
math score                       81
reading score                    72
writing score                    77
dtype: int64
------------------------------------------------------
Total de filas duplicadas
0
------------------------------------------------------
Nombres duplicados


Empty DataFrame
Columns: [names, gender, race/ethnicity, parental level of education, lunch, test preparation course, math score, reading score, writing score]
Index: []

**Justificación:**

Al eliminar primero las filas con notas fuera de rango, tengo el respaldo de que alguna de las otras filas con el mismo nombre asociado pueda tener notas correctamente ingresadas. En caso contrario, podría quedarme con la primera aparición de un alumno en el dataset y que esta tenga notas mal ingresadas, luego al verificar el rango de las notas, podría eliminar al alumno de la base de datos. 

## 3. Trabajando con Datos [2.0 puntos]

<p align="center">
  <img src="https://media3.giphy.com/media/YQitE4YNQNahy/giphy-downsized-large.gif">
</p>

### 3.1 Transformación de notas a "Nota Chilena" [0.3 Puntos]

En una de las reuniones con Don Caguayo este les comenta su intención de pasar todas sus notas a la "Escala Chilena". Para esto les facilita la siguiente ecuación:

$$Nota \, Chilena= \% \, de \, logro*0.06 + 1$$

Utilizando la ecuación entregada, cree una función que le permita pasar a la escala Chilena las notas del dataframe y entregue las notas redondeadas al segundo decimal.

**To-Do**

- [ ] Tranforme los valores a la escala chilena.

**Respuesta:**

In [15]:
# Escriba su código aquí
def nota_chilena(dataframe_in, columna):
  dataframe_in[columna] = dataframe_in[columna].apply(lambda x: np.round(x*0.06 + 1.0, 2))
  return dataframe_in

In [16]:
cols_score =  df_grades.filter(like='score').columns

for col in cols_score:
    nota_chilena(df_grades, col)

In [17]:
#Muestre sus resultados aquí
df_grades

names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
3     Charles Linstrom    male        group A           bachelor's degree   
5          Brian Young    male        group C            some high school   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
...                ...     ...            ...                         ...   
1392     Richard Young    male        group D                 high school   
1394     Wanda Russell  female        group B                 high school   
1396    Marina Zeigler  female        group C           bachelor's degree   
1397     Laurie Carter  female        group B            some high school   
1399      Amanda Perez  female        group A                 high school   

             lunch test preparation course  math score  reading score  \
0         standard                    none        3.22           3.76   
3         standard               completed        5.80           5.68   
5         standard                    none        5.38           4.96   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
...            ...                     ...         ...            ...   
1392      standard                    none        5.14           5.50   
1394  free/reduced               completed        2.38           3.64   
1396  free/reduced               completed        4.96           5.44   
1397      standard               completed        4.24           4.66   
1399      standard               completed        5.08           5.80   

      writing score  
0              3.76  
3              5.86  
5              4.78  
6              5.56  
7              6.22  
...             ...  
1392           5.26  
1394           3.16  
1396           5.86  
1397           4.72  
1399           5.56  

[1000 rows x 9 columns]

### 3.2 Obtención de promedio de alumnos [0.3 Puntos]

Con todas las notas adjuntadas, la prestigiosa Universidad De la Cachaña necesita obtener el promedio general de cada uno de los alumnos. Para esto se le solicita que calcule el promedio con cada una de las asignaturas y el promedio se anexe como una nueva columna al final del `DataFrame` original. Cabe señalar que las notas deben ser redondeadas al segundo decimal.

**To-Do:**
- [ ] Crear función que nos permite obtener el promedio de cada uno de los alumnos y anexar esta información en el `DataFrame` original.


**Respuesta:**

In [18]:
#Escriba su código aquí
def alumnos_promedio(dataframe_in):
  cols_score =  dataframe_in.filter(like='score').columns
  dataframe_in['Promedio'] = dataframe_in[cols_score].sum(axis=1)/3
  return dataframe_in

In [19]:
#Muestre sus resultados aquí
df_grades = alumnos_promedio(df_grades)
df_grades.head()

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
3  Charles Linstrom    male        group A           bachelor's degree   
5       Brian Young    male        group C            some high school   
6    Howard Jimenez    male        group E            some high school   
7      Wayne Wilson    male        group B            some high school   

      lunch test preparation course  math score  reading score  writing score  \
0  standard                    none        3.22           3.76           3.76   
3  standard               completed        5.80           5.68           5.86   
5  standard                    none        5.38           4.96           4.78   
6  standard               completed        5.86           5.50           5.56   
7  standard               completed        6.64           6.16           6.22   

   Promedio  
0      3.58  
3      5.78  
5      5.04  
6      5.64  
7      6.34

### 3.3 Reprobados y Destacados [0.9 Puntos, 0.3 cada DataFrame]

Contentos con su desempeño, la Universidad les solicita que obtengan los alumnos quienes han reprobado el año y los que han tenido desempeño de destacado. Para esto se considerará que un alumno ha reprobado el año cuando el promedio de notas es inferior a 4.0 en la Escala Chilena, y se considerará destacado si posee notas superiores a 5.8 en todas las asignaturas (o sea notas superiores a 5.8 en `math score`, `reading score` y `writing score`).

Cree una función que permita obtener tres Dataframes:
- El primero deberá poseer solamente los alumnos que han reprobado en la prestigiosa universidad.
- El segundo debe entregar un DataFrame similar al original, pero marcado con una "R" el promedio de notas de los alumnos reprobados.
- El tercero será un Data Frame con los alumnos destacados ordenados de mayor a menor.

**To-Do:**

- [ ] Obtener DataFrame con los alumnos que reprobaron el año.
- [ ] Obtener DataFrame similar al original, pero marcados con una "R" el atributo GPA de los alumnos reprobados.
- [ ] Obtener DataFrame con los alumnos destacados ordenados de mayor a menor.

**Respuesta:**

In [26]:
df_grades.loc[(df_grades['math score']>5.8)|(df_grades['reading score']>5.8)|(df_grades['writing score']>5.8)]

names  gender race/ethnicity parental level of education  \
3     Charles Linstrom    male        group A           bachelor's degree   
6       Howard Jimenez    male        group E            some high school   
7         Wayne Wilson    male        group B            some high school   
16         Damon Mckee    male        group E                 high school   
17        Sophia Bijou  female        group D            some high school   
...                ...     ...            ...                         ...   
1375       Linda Nevin  female        group B            some high school   
1377  Timothy Anderson    male        group C          associate's degree   
1380    Brandon Palmer    male        group D          associate's degree   
1381   Robert Morrison    male        group D                some college   
1396    Marina Zeigler  female        group C           bachelor's degree   

             lunch test preparation course  math score  reading score  \
3         standard               completed        5.80           5.68   
6         standard               completed        5.86           5.50   
7         standard               completed        6.64           6.16   
16        standard                    none        6.64           5.38   
17        standard               completed        4.90           5.68   
...            ...                     ...         ...            ...   
1375      standard               completed        4.90           5.92   
1377  free/reduced               completed        5.68           5.86   
1380  free/reduced                    none        6.40           6.22   
1381      standard               completed        7.00           6.82   
1396  free/reduced               completed        4.96           5.44   

      writing score  Promedio  
3              5.86      5.78  
6              5.56      5.64  
7              6.22      6.34  
16             5.26      5.76  
17             5.92      5.50  
...             ...       ...  
1375           5.68      5.50  
1377           5.92      5.82  
1380           5.50      6.04  
1381           6.94      6.92  
1396           5.86      5.42  

[299 rows x 10 columns]

In [36]:
#Escriba su código aquí
def reprobados_destacados(dataframe_in):
    # retorna 3 df
    reprobados = dataframe_in.loc[dataframe_in['Promedio']< 4.0]

    reprobados_r = reprobados.copy()
    reprobados_r['Promedio'] = 'R'

    # notas > 5.8 en todas las asignaturas
    destacados = dataframe_in.loc[(dataframe_in['math score']>5.8)|
                                  (dataframe_in['reading score']>5.8)|
                                  (dataframe_in['writing score']>5.8)]
    destacados = destacados.sort_values('Promedio', ascending=False)
    return reprobados, reprobados_r, destacados


In [37]:
# Muestre sus resultados aquí
dataframe1, dataframe2, dataframe3 = reprobados_destacados(df_grades)

In [38]:
display(dataframe1.head())
display(dataframe2.head())
display(dataframe3.head())

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
13      Mikki Amaya  female        group A                 high school   
26    Samuel Dougal    male        group C             master's degree   
55  Rodger Talavera    male        group B                 high school   
63   Morris Johnson    male        group A          associate's degree   

           lunch test preparation course  math score  reading score  \
0       standard                    none        3.22           3.76   
13  free/reduced               completed        3.04           3.88   
26  free/reduced               completed        3.76           3.52   
55  free/reduced                    none        3.16           2.74   
63  free/reduced               completed        3.40           4.30   

    writing score  Promedio  
0            3.76      3.58  
13           3.46      3.46  
26           3.76      3.68  
55           2.62      2.84  
63           4.18      3.96

names  gender race/ethnicity parental level of education  \
0     Rita Courtney  female        group B            some high school   
13      Mikki Amaya  female        group A                 high school   
26    Samuel Dougal    male        group C             master's degree   
55  Rodger Talavera    male        group B                 high school   
63   Morris Johnson    male        group A          associate's degree   

           lunch test preparation course  math score  reading score  \
0       standard                    none        3.22           3.76   
13  free/reduced               completed        3.04           3.88   
26  free/reduced               completed        3.76           3.52   
55  free/reduced                    none        3.16           2.74   
63  free/reduced               completed        3.40           4.30   

    writing score Promedio  
0            3.76        R  
13           3.46        R  
26           3.76        R  
55           2.62        R  
63           4.18        R

names  gender race/ethnicity parental level of education  \
384   Elizabeth Beasley  female        group E           bachelor's degree   
1352      Sandra Pompey  female        group E          associate's degree   
677       Richard Pryor    male        group E           bachelor's degree   
475           Emma Gray  female        group E           bachelor's degree   
1276     Angela Cressey  female        group D                some college   

         lunch test preparation course  math score  reading score  \
384   standard                    none        7.00            7.0   
1352  standard                    none        7.00            7.0   
677   standard               completed        7.00            7.0   
475   standard               completed        6.94            7.0   
1276  standard                    none        6.88            7.0   

      writing score  Promedio  
384            7.00      7.00  
1352           7.00      7.00  
677            7.00      7.00  
475            7.00      6.98  
1276           6.94      6.94

### 3.4 Quintiles de Rendimiento [0.5 Puntos]

El rector, intrigado por saber si existe una relación entre el nivel educacional de los padres y las notas de sus hijos les plantea estudiar dicha relación. Para esto, les solicita que desde el `DataFrame` que posee a todos los alumnos (`DataFrame` obtenido tras la limpieza), generen una nueva columna que clasifique en quintiles de rendimiento para sus alumnos: `['Bajísimo', 'Bajo', 'Medio', 'Alto', 'Muy alto']`. (investigar `qcut()`)

Hecho esto, grafique a través de un scatter plot el nivel educacional de los padres v/s el promedio de los alumnos, utilice los cuantiles generados anteriormente para mostrar en que cuantil se encuentra cada uno de los datos.

Realizado el grafico, responda la siguiente pregunta: **¿Es posible visualizar alguna tendencia en el gráfico?.**



**Respuesta:**

In [ ]:
# Escriba su código para quintiles aquí

"""
NOTA: La variable del dataframe que contenga los quintiles debe llamarse
'GPA cuantiles' para que el gráfico funcione.
"""

In [ ]:
# Código de gráfico del amigo ChatGPT
fig = px.scatter(
    df_grades.sort_values(by="GPA"),
    x="GPA",
    y="parental level of education",
    hover_data=["gender"],
    color="GPA cuantiles",
    title="GPA por Nivel de educación de los padres",
    color_discrete_sequence=["tomato", "orange", "yellow", "greenyellow", "green"]
)

fig

```
Escriba aqui su respuesta
```

## 4. Operaciones con tablas [2.0 puntos]

<p align="center">
  <img src="https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExYWx5dTBkMGhhOTJmeGQzbDV5M21rNTN1NWVwMm1qbHd3cjh3Nmk4NSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/13HBDT4QSTpveU/giphy.gif">
</p>



### 4.1 Carga de datos y concatenar dataset [0.2 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente).

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: `students_grades_1.json` y `students_grades_2.json`.


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos.

**No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas**, recuerde que anteriormente ya se encargo de este tema.


Hint: investigue el método `concat`

**Respuesta:**

In [ ]:
# Inserte su respuesta aquí
df_grades_1= ...
df_grades_2= ...

df_grades= ...

### 4.2 Análisis por Nivel Educacional Etnia de los Padres [0.8 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos.
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.


**Tabla de Resumen [0.5 Puntos]**


Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar el método `map()`) de la variable `parental level of education`  según la siguiente conversión:

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas.
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'.

Utilizar la tabla de resultados esperados como guía para desarrollar este punto.


**Análisis [0.3 Puntos]**

Finalmente, responda las siguientes preguntas sobre el gráfico de caja que viene implementado a continuación. En este se muestra el GPA con respecto al nivel educacional, donde la variable de color es la etnicidad.

- [ ]  ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?


**Respuesta:**

In [ ]:
# Inserte su respuesta aquí

In [ ]:
fig = px.box(
    df_grades,
    x="simple parental level of education",
    y="GPA",
    color="race/ethnicity",
    hover_data=["GPA"],
)
fig.show()

    Análizar gráfica acá.

**Resultados Esperados**

Esta tabla ilustra la estructura del dataframe a conseguir. Los valores son referenciales, no necesariamente son los esperados.

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.83 |            5.05 |            4.95 |  4.94 |      26 | 2.6 %        |
|  1 |                  | postgraduate                         |         4.46 |            4.88 |            5.06 |  4.80 |       3 | 0.3 %        |
|  2 |                  | school                               |         4.65 |            4.81 |            4.66 |  4.71 |      60 | 6.0 %        |
|  3 | group B          | college                              |         5.03 |            5.24 |            5.16 |  5.14 |      61 | 6.1 %        |
|  4 |                  | postgraduate                         |         5.03 |            5.81 |            5.63 |  5.49 |       6 | 0.6 %        |
|  5 |                  | school                               |         4.69 |            4.90 |            4.79 |  4.79 |     123 | 12.3 %       |
|  6 | group C          | college                              |         5.03 |            5.36 |            5.33 |  5.24 |     118 | 11.8 %       |
|  7 |                  | postgraduate                         |         5.02 |            5.23 |            5.17 |  5.14 |      19 | 1.9 %        |
|  8 |                  | school                               |         4.74 |            5.00 |            4.89 |  4.88 |     182 | 18.2 %       |
|  9 | group D          | college                              |         5.06 |            5.22 |            5.24 |  5.17 |      78 | 7.8 %        |
| 10 |                  | postgraduate                         |         5.35 |            5.63 |            5.78 |  5.59 |      23 | 2.3 %        |
| 11 |                  | school                               |         4.99 |            5.13 |            5.11 |  5.08 |     161 | 16.1 %       |
| 12 | group E          | college                              |         5.53 |            5.45 |            5.43 |  5.47 |      57 | 5.7 %        |
| 13 |                  | postgraduate                         |         5.48 |            5.93 |            5.83 |  5.74 |       8 | 0.8 %        |
| 14 |                  | school                               |         5.35 |            5.27 |            5.11 |  5.25 |      75 | 7.5 %        |


### 4.3 Joins, joins y más joins [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias.
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene registros de alumnos que ya no forman parte de la universidad, por lo que no deben ser parte de este análisis.


El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`).


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados o **alumnos que no pertenecen a la universidad**).

**To-Do**


- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

**Respuesta:**

In [ ]:
# Inserte su respuesta aquí
other_grades = ...

df_grades_outer = ...
df_grades_left = ...
df_grades_right = ...
df_grades_inner = ...

        Explique sus resultados acá

# Conclusión


Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>